includes/express-checkout/class-wc-payments-express-checkout-ajax-handler.php

```diff
@@ -285,15 +285,142 @@ private function transform_ece_address_state_data( $address ) {
 			return $address;
 		}
 
-		// states from Apple Pay or Google Pay might be in long format, we need their short format.
+		// Due to a bug in Apple Pay, the "Region" part of a Hong Kong address is delivered in
+		// `shipping_postcode`, so we need some special case handling for that. According to
+		// our sources at Apple Pay people will sometimes use the district or even sub-district
+		// for this value. As such we check against all regions, districts, and sub-districts
+		// with both English and Mandarin spelling.
+		//
+		// @reykjalin: The check here is quite elaborate in an attempt to make sure this doesn't break once
+		// Apple Pay fixes the bug that causes address values to be in the wrong place. Because of that the
+		// algorithm becomes:
+		// 1. Use the supplied state if it's valid (in case Apple Pay bug is fixed)
+		// 2. Use the value supplied in the postcode if it's a valid HK region (equivalent to a WC state).
+		// 3. Fall back to the value supplied in the state. This will likely cause a validation error, in
+		// which case a merchant can reach out to us so we can either: 1) add whatever the customer used
+		// as a state to our list of valid states; or 2) let them know the customer must spell the state
+		// in some way that matches our list of valid states.
+		//
+		// @reykjalin: This HK specific sanitazation *should be removed* once Apple Pay fix
+		// the address bug. More info on that in pc4etw-bY-p2.
+		if ( Country_Code::HONG_KONG === $country ) {
+			include_once WCPAY_ABSPATH . 'includes/constants/class-express-checkout-hong-kong-states.php';
+
+			$state = $address['state'] ?? '';
+			if ( ! empty( $state ) && ! \WCPay\Constants\Express_Checkout_Hong_Kong_States::is_valid_state( strtolower( $state ) ) ) {
+				$postcode = $address['postcode'] ?? '';
+				if ( ! empty( $postcode ) && \WCPay\Constants\Express_Checkout_Hong_Kong_States::is_valid_state( strtolower( $postcode ) ) ) {
+					$address['state'] = $postcode;
+				}
+			}
+		}
+
+		// States from Apple Pay or Google Pay are in long format, we need their short format.
 		$state = $address['state'] ?? '';
 		if ( ! empty( $state ) ) {
-			$address['state'] = $this->express_checkout_button_helper->get_normalized_state( $state, $country );
+			$address['state'] = $this->get_normalized_state( $state, $country );
```

Comment:
I colocated the `get_normalized_state` method and its descendants ( `is_normalized_state` / `get_normalized_state_from_ece_states` / `get_normalized_state_from_wc_states` ) from the `express_checkout_button_helper ` helper to this class, since they were only used here.